File: mobile/expo-driver-app/src/services/auth.service.ts
import apiService from './api.service';
import { API_ENDPOINTS } from '../config/api';
import { saveToken, saveUser, saveDriver, clearAuth, saveProfile } from './storage.service';

export interface LoginRequest {
  email: string;
  password: string;
}

export interface LoginResponse {
  success: boolean;
  token?: string;
  user?: any;
  driver?: any;
  error?: string;
}

class AuthService {
  async login(email: string, password: string): Promise<LoginResponse> {
    try {
      const response = await apiService.post<LoginResponse>(API_ENDPOINTS.LOGIN, {
        email,
        password,
      });

      if (response.success && response.token) {
        console.log('🔑 Saving token:', response.token);
        await saveToken(response.token);
        
        // ✅ CRITICAL: Ensure user object includes driver relation
        if (response.user) {
          console.log('💾 Saving user with driver relation:', {
            userId: response.user.id,
            driverId: response.user.driver?.id,
            hasDriverRelation: !!response.user.driver
          });
          
          await saveUser(response.user);
          
          // Save driver separately for backward compatibility
          if (response.user.driver) {
            await saveDriver(response.user.driver);
          } else if (response.driver) {
            // Fallback: if driver is separate in response
            await saveDriver(response.driver);
            // Update user object to include driver relation
            response.user.driver = response.driver;
            await saveUser(response.user);
          }
        }
        
        console.log('✅ Token and user data saved successfully');
      } else {
        console.log('❌ No token in response:', response);
      }

      return response;
    } catch (error: any) {
      console.error('Login error:', error);
      throw error;
    }
  }

  async logout(): Promise<void> {
    await clearAuth();
  }

  async checkSession(): Promise<any> {
    try {
      return await apiService.get(API_ENDPOINTS.SESSION);
    } catch (error) {
      console.error('Session check error:', error);
      throw error;
    }
  }
}

export default new AuthService();

File: mobile/ios-driver-app/Services/AuthService.swift
import Foundation

// MARK: - Authentication Service

class AuthService {
    static let shared = AuthService()
    private let network = NetworkService.shared
    
    private init() {}
    
    // MARK: - Login
    
    func login(email: String, password: String) async throws -> LoginResponse {
        let request = LoginRequest(email: email, password: password)
        
        let response: LoginResponse = try await network.request(
            .login,
            method: .post,
            body: request,
            requiresAuth: false
        )
        
        // Check for errors
        if let error = response.error {
            throw AuthError.loginFailed(error)
        }
        
        // Save token and user data
        if let token = response.token {
            TokenStorage.saveToken(token)
        }
        
        if let user = response.user {
            TokenStorage.saveUser(user)
        }
        
        if let driver = response.driver {
            TokenStorage.saveDriver(driver)
        }
        
        return response
    }
    
    // MARK: - Logout
    
    func logout() async throws {
        // Clear local storage
        TokenStorage.clearAll()
        
        // Optionally call logout endpoint
        // let _: EmptyResponse = try await network.request(.logout, method: .post)
    }
    
    // MARK: - Session Check
    
    func checkSession() async throws -> SessionResponse {
        let response: SessionResponse = try await network.request(.session)
        
        // Update stored user data
        if let user = response.user {
            TokenStorage.saveUser(user)
        }
        
        if let driver = response.driver {
            TokenStorage.saveDriver(driver)
        }
        
        return response
    }
    
    // MARK: - Get Profile
    
    func getProfile() async throws -> DriverProfileResponse {
        return try await network.request(.profile)
    }
    
    // MARK: - Check Stored Session
    
    func hasStoredSession() -> Bool {
        return TokenStorage.getToken() != nil
    }
    
    func getStoredUser() -> User? {
        return TokenStorage.getUser()
    }
    
    func getStoredDriver() -> Driver? {
        return TokenStorage.getDriver()
    }
}

// MARK: - Auth Error

enum AuthError: LocalizedError {
    case loginFailed(String)
    case sessionExpired
    case notAuthenticated
    case notApproved
    
    var errorDescription: String? {
        switch self {
        case .loginFailed(let message):
            return message
        case .sessionExpired:
            return "Your session has expired. Please login again."
        case .notAuthenticated:
            return "You are not authenticated. Please login."
        case .notApproved:
            return "Your account is not yet approved. Please wait for admin approval."
        }
    }
}

// MARK: - Empty Response

struct EmptyResponse: Codable {}

